File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode *root) {
        if (!root) {
            return 0;
        }
        int l = solve(root->left);
        int r = solve(root->right);
        if (l == -1 || r == -1 || abs(l - r) > 1) {
            return -1;
        }
        return 1 + max(l, r);
    }
    bool isBalanced(TreeNode* root) {
        return solve(root) != -1;
    }
};

File: LeetCode/0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def checkHeightAndBalance(node):
            if not node:
                return 0
            
            left_height = checkHeightAndBalance(node.left)
            right_height = checkHeightAndBalance(node.right)
            
            if left_height == -1 or right_height == -1 or abs(left_height - right_height) > 1:
                return -1
            
            return 1 + max(left_height, right_height)
        
        return checkHeightAndBalance(root) != -1